XML_to_SON/grammar_parser_test.py

```diff
@@ -0,0 +1,305 @@
+import xml.etree.ElementTree as ET
+import re
+import pprint
+import json
+import os.path
+import argparse
+
+def get_element_type(node):
+    for child in node:
+        if child.tag.endswith("data") and "type" in child.attrib:
+            xml_type = child.attrib['type']
+            return conversion_dict.get(xml_type, 'Unknown')
+    return None
+
+def process_element(node, parent_attrib=None):
+    if parent_attrib is None:
+        parent_attrib = {}
+    ele_dict = {}
+    name = node.attrib.get('name') 
+    ele_dict[name] = {}
+
+    if not parent_attrib:
+        ele_dict[name]['minOccurs'] = 1
+        ele_dict[name]['maxOccurs'] = 1
+
+    element_type = get_element_type(node)
+    if element_type:
+        ele_dict[name]['ValType'] = element_type
+    else:
+        for child in node:
+            if child.tag.endswith("text"):
+                ele_dict[name]['ValType'] = "String"
+                break
+
+    for child in node:
+        child_attrib = {}
+        if child.tag.endswith("choice"):
+            choice_str = handle_choice_element(child)
+            if choice_str:
+                ele_dict[name]['ChildExactlyOne'] = f"{choice_str}"
+            break
+        if child.tag not in {f"{ns}zeroOrMore", f"{ns}oneOrMore", f"{ns}optional"}:
+            child_attrib = {'minOccurs': 1, 'maxOccurs': 1}
+        ele_dict[name].update(process_node(child, child_attrib))
+
+    ele_dict[name].update(parent_attrib)
+
+    return ele_dict
+
+def process_node(node, add_attrib={}):
+            
+    node_dict = {}
+    if node.tag.endswith("element"):
+        node_dict.update(process_element(node, add_attrib))
+    elif node.tag.endswith("optional"):
+        for child in node:
+            node_dict.update(process_node(child, {'minOccurs': 0, 'maxOccurs': 1}))
+    elif node.tag.endswith("zeroOrMore"):
+        for child in node:
+            node_dict.update(process_node(child, {'minOccurs': 0}))
+    elif node.tag.endswith("oneOrMore"):
+        for child in node:
+            node_dict.update(process_node(child, {'minOccurs': 1}))
+    else:
+        for child in node:
+            node_dict.update(process_node(child))
```

Comment:
If possible, repeat as little as possible
```suggestion
    attrib_dict = {}
    if node.tag.endswith("element"):
        node_dict.update(process_element(node, add_attrib))
    elif node.tag.endswith("optional"):
        attrib_dict = {'minOccurs': 0, 'maxOccurs': 1}
    elif node.tag.endswith("zeroOrMore"):
        attrib_dict = {'minOccurs': 0}
    elif node.tag.endswith("oneOrMore"):
        attrib_dict = {'minOccurs': 1}

    for child in node:
        node_dict.update(process_node(child, attrib_dict))
```

Reply:
This change does not work. I get a key error later on while integrating the detailed schemas from the metadata. I believe this happens because this suggestion ignores the for loop of the last else. The last else is the one that allows for the "choice" that later gets processed as ChildExactlyOne.

Reply:
I could probably make a new function with the for loop that takes the attrib_dict as a parameter and call the function inside every elif to avoid having the for loop every time.

Reply:
no - you can ignore my suggestion.  I think it doesn't work because it calls the for loop even in the case when the first if statement matches, which is wrong - my bad.